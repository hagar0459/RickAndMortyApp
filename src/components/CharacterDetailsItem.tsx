/**
 * @flow
 * Created by Hagar Abdelghafar on 04.07.2020
 */
import React, { FC } from 'react';
import { Image, StyleSheet, Text, View } from 'react-native';
import type { CharcterDetailsType } from '../graphql/queries/requests';


type Props = {
  info: CharcterDetailsType | undefined;
};


export const CharacterDetailsItem: FC<Props> = ({ info }: Props) => {

  const renderInfoContainer = () => (
    <View style={styles.infoContainer}>
      <Text style={styles.name}>{info?.character.name}</Text>
      <Text style={styles.gender}>{info?.character.gender}</Text>


      <Text style={styles.species}>{info?.character.species}</Text>

    </View>
  );
  const renderImage = () => {
    return (
      <Image
        style={styles.image}
        source={{ uri: info?.character.image }}
        resizeMode={'cover'}
      />
    );
  };

  const renderContent = () => {

    return (
      <>
        {renderImage()}
        {renderInfoContainer()}

      </>
    );


  };


  return (
    <View style={styles.card} >
      {renderContent()}
    </View>
  );
}

const styles = StyleSheet.create({

  card: {
    borderRadius: 0,
    marginBottom: 20

  },
  image: {
    height: 150,

  },
  infoContainer: {
    padding: 16,
  },
  name: {
    fontSize: 22,
    fontWeight: 'bold',
  },
  gender: {
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 8,
  },
  species: {
    fontSize: 16,
    fontWeight: '400',
    color: '#787878',
  },
});
