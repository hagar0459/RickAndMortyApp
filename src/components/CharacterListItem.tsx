/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import React, { FC } from 'react';
import { Dimensions, Image, StyleSheet, Text, View, TouchableOpacity } from 'react-native';
import type { CharcterResultType } from '../graphql/queries/requests';


type Props = {
  info: CharcterResultType;
  onPress(item: CharcterResultType): void;
  index: number;
};
const ITEM_HEIGHT = 80;

export const CharacterListItem: FC<Props> = ({ info, onPress, index }: Props) => {

  const renderText = () => (
    <View  >
      <Text style={styles.title}>{info.name}</Text>
    </View>
  );
  const renderImage = (isRight: boolean) => {
    const rawUri = info.image;
    if (!rawUri) {
      return null;
    }
    const uri = `${rawUri ?? ''}?width=300`;
    // console.log({ uri });
    return (
      <View style={isRight ? styles.right : styles.left}>
        <Image source={{ uri }} resizeMode="cover" style={styles.image} />
      </View>
    );
  };

  const renderContent = () => {
    const isRight = index % 2 === 0;
    if (isRight) {
      return (
        <>
          {renderText()}
          {renderImage(isRight)}
        </>
      );
    }
    return (
      <>
        {renderImage(isRight)}
        {renderText()}
      </>
    );
  };


  return (
    <TouchableOpacity style={styles.card}

      onPress={() => {
        onPress(info);
      }}
    >

      {renderContent()}
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  card: {
    backgroundColor: 'white',
    borderRadius: 16,
    shadowOpacity: 0.2,
    shadowRadius: 4,
    shadowColor: 'black',
    shadowOffset: {
      height: 0,
      width: 0,
    },
    elevation: 1,
    marginVertical: 10,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',


  },
  image: {
    height: ITEM_HEIGHT,
    width: ITEM_HEIGHT,
  },
  left: {
    borderTopLeftRadius: 8,
    borderBottomLeftRadius: 8,
    overflow: 'hidden',

  },
  right: {
    borderTopRightRadius: 8,
    borderBottomRightRadius: 8,
    overflow: 'hidden',

  },
  title: {
    fontSize: 17,
    padding: 5,
    width: Dimensions.get('window').width - ITEM_HEIGHT - 30,
    textAlign: 'center',
  },
});
