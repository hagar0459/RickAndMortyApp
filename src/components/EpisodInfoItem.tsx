/**
 * @flow
 * Created by Hagar Abdelghafar on 05.06.2021
 */
import React, { FC } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import type { EpisodeType } from '../graphql/queries/requests';


type Props = {
  info: EpisodeType;

};

export const EpisodInfoItem: FC<Props> = ({ info }: Props) => {

  return (
    <View style={styles.card}  >
      <Text style={styles.name}>{info.name}</Text>
      <Text style={styles.air_date}>{info.air_date}</Text>
    </View>



  );
}

const styles = StyleSheet.create({
  card: {
    backgroundColor: 'white', margin: '5%', borderRadius: 10, width: '40%', padding: 10, alignContent: 'center', alignItems: 'center'
  },
  name: {
    fontSize: 15,
    fontWeight: '400',
    marginBottom: 8,
  },
  air_date: {
    fontSize: 12,
    fontWeight: '200',
    color: '#787878',
  },
});
