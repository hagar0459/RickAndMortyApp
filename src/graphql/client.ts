/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import { ApolloClient, InMemoryCache } from "@apollo/client";
import { GRAPHQL_URL } from "../config";

export const apolloClient = new ApolloClient({
  uri: GRAPHQL_URL,
  cache: new InMemoryCache({}),
  defaultOptions: {
    watchQuery: {
      fetchPolicy: "cache-first",
    },
  },
});
