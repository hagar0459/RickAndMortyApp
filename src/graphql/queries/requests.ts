/**
 * @flow
 * Created by Hagar Abdelghafar on 04.06.2021
 */
import { gql } from '@apollo/client';
export const SEARCH_CHARACTER = gql`
  query SearchCharacter($page: Int!, $name: String!) {
    characters(page: $page, filter: { name: $name }) {
      results {
        name
        image
        id
      }
      info {
        next
        count
      }
    }
  }
`;
export const GET_CHARACTER_DETAILS = gql`
  query GetCharacterDetails($id: ID!) {
    character(id: $id) {
      name
      id
      image
      gender
      species
      episode {
        air_date
        name
      }
    }
  }
`;
export type CharcterListType = {
  characters: {
    info: characterListInfoType;
    results: Array<CharcterResultType>;
  };
};
export type CharcterResultType = {
  id: number;
  name: string;
  image: string;
};
export type characterListInfoType = {
  next: number | null;
  count: number;
};
export type CharcterDetailsType = {
  character: {
    id: number;
    name: string;
    image: string;
    gender: string;
    species: string;
    episode: Array<EpisodeType>;
  };
};
export type EpisodeType = {
  air_date: string;
  name: string;
};
