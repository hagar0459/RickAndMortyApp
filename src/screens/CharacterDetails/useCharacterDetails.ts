/**
 * @flow
 * Created by Hagar Abdelghafar on 05.06.2021
 */
import React, { useState, useEffect, FC } from "react";
import { useLazyQuery } from "@apollo/client";
import {
  GET_CHARACTER_DETAILS,
  CharcterDetailsType,
} from "../../graphql/queries/requests";

type Props = {
  characterId: number;
};

type Result = {
  data: CharcterDetailsType | undefined;
  getCharacterDetails(): void;
  loading: boolean;
  error: string | undefined;
};

export const useCharacterDetails = ({ characterId }: Props): Result => {
  const [getCharacterDetails, { data, error, loading }] =
    useLazyQuery<CharcterDetailsType>(GET_CHARACTER_DETAILS, {
      fetchPolicy: "cache-and-network",
      nextFetchPolicy: "cache-first",
      variables: {
        id: characterId,
      },
    });

  useEffect(() => {
    if (!loading) {
      getCharacterDetails();
    }
  }, []);

  return {
    data,
    getCharacterDetails,
    loading,
    error: error?.message,
  };
};
