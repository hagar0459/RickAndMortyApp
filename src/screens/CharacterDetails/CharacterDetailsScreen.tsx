/**
 * @flow
 * Created by Hagar Abdelghafar on 05.06.2021
 */
import React, { FC } from 'react';
import
{
  StyleSheet,
  FlatList,
  ListRenderItemInfo,
  Text,
  Alert,
  SafeAreaView,
} from 'react-native';
import { CharacterDetailsItem } from '../../components/CharacterDetailsItem';
import { EpisodInfoItem } from '../../components/EpisodInfoItem';
import { EpisodeType } from '../../graphql/queries/requests';
import { useRoute, RouteProp } from '@react-navigation/native';
import { useCharacterDetails } from './useCharacterDetails';
import { RootStackParamList } from './../../Navigation'
type ProfileScreenRouteProp = RouteProp<RootStackParamList, 'NAVIGATION_CHARACTERS_DETAILS_ROUTE'>;

export const CharacterDetailsScreen: FC = () =>
{
  const route = useRoute<ProfileScreenRouteProp>();
  const { data, error } = useCharacterDetails( {
    characterId: route?.params?.characterId
  } );
  if ( error )
  {
    Alert.alert( error );
  }
  const renderEpisod = ( { item }: ListRenderItemInfo<EpisodeType> ) =>
  {
    return <EpisodInfoItem info={item} />;
  };

  return (
    <SafeAreaView>
      <CharacterDetailsItem info={data}></CharacterDetailsItem>
      {!error && (
        <Text style={{ marginHorizontal: 20, fontSize: 16, fontWeight: '600' }}>
          Episods:
        </Text>
      )}
      <FlatList
        numColumns={2}
        data={data?.character.episode}
        keyExtractor={( { name } ) => name}
        renderItem={renderEpisod}
      />
    </SafeAreaView>
  );
};
