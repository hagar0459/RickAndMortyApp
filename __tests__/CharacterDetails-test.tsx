import * as React from 'react';
import {CharacterDetailsScreen  } from '../src/screens/CharacterDetails/CharacterDetailsScreen';
import renderer from 'react-test-renderer';
import {MockedProvider} from '@apollo/client/testing';
import {GET_CHARACTER_DETAILS} from '../src/graphql/queries/requests';
import {FlatList ,Image} from 'react-native';
import{CharacterDetailsItem} from '../src/components/CharacterDetailsItem'
import{EpisodInfoItem} from '../src/components/EpisodInfoItem'

async function wait(ms = 0) {
  await renderer.act(() => {
    return new Promise(resolve => {
      setTimeout(resolve, ms);
    });
  });
}

describe('Testing CharacterDetailsScreen', () => {

it('Screen Shows FlatList', () => {
    const component = renderer.create(
      <MockedProvider>
        <CharacterDetailsScreen />
      </MockedProvider>,
    );

    expect(() => component.root.findByType(FlatList)).not.toThrow();
  });

it('Screen Shows Image', () => {
    const component = renderer.create(
      <MockedProvider>
        <CharacterDetailsScreen />
      </MockedProvider>,
    );

    expect(() => component.root.findByType(Image)).not.toThrow();
  });

it('Screen Shows EpisodInfoItem', () => {
    const component = renderer.create(
      <MockedProvider>
        <CharacterDetailsScreen />
      </MockedProvider>,
    );

    expect(() => component.root.findByType(EpisodInfoItem)).not.toThrow();
  });

it('Screen Shows CharacterDetailsItem', () => {
    const component = renderer.create(
      <MockedProvider>
        <CharacterDetailsScreen />
      </MockedProvider>,
    );

    expect(() => component.root.findByType(CharacterDetailsItem)).not.toThrow();
  });

it('Shows Charcter Details', async () => {
    const mockedData = {
      request: {
        query: GET_CHARACTER_DETAILS,
        variables: { 
          id: 1}
      },
      result: {
        "data": {
          "character": {
            "name": "Rick Sanchez",
            "id": "1",
            "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
            "gender": "Male",
            "species": "Human",
            "episode": [
              {
                "air_date": "December 2, 2013",
                "name": "Pilot"
              },
              {
                "air_date": "December 9, 2013",
                "name": "Lawnmower Dog"
              },
              {
                "air_date": "December 16, 2013",
                "name": "Anatomy Park"
              },
              {
                "air_date": "January 13, 2014",
                "name": "M. Night Shaym-Aliens!"
              },
              {
                "air_date": "January 20, 2014",
                "name": "Meeseeks and Destroy"
              },
              {
                "air_date": "January 27, 2014",
                "name": "Rick Potion #9"
              },
              {
                "air_date": "March 10, 2014",
                "name": "Raising Gazorpazorp"
              },
              {
                "air_date": "March 17, 2014",
                "name": "Rixty Minutes"
              },
              {
                "air_date": "March 24, 2014",
                "name": "Something Ricked This Way Comes"
              },
              {
                "air_date": "April 7, 2014",
                "name": "Close Rick-counters of the Rick Kind"
              },
              {
                "air_date": "April 14, 2014",
                "name": "Ricksy Business"
              },
              {
                "air_date": "July 26, 2015",
                "name": "A Rickle in Time"
              },
              {
                "air_date": "August 2, 2015",
                "name": "Mortynight Run"
              },
              {
                "air_date": "August 9, 2015",
                "name": "Auto Erotic Assimilation"
              },
              {
                "air_date": "August 16, 2015",
                "name": "Total Rickall"
              },
              {
                "air_date": "August 23, 2015",
                "name": "Get Schwifty"
              },
              {
                "air_date": "August 30, 2015",
                "name": "The Ricks Must Be Crazy"
              },
              {
                "air_date": "September 13, 2015",
                "name": "Big Trouble in Little Sanchez"
              },
              {
                "air_date": "September 20, 2015",
                "name": "Interdimensional Cable 2: Tempting Fate"
              },
              {
                "air_date": "September 27, 2015",
                "name": "Look Who's Purging Now"
              },
              {
                "air_date": "October 4, 2015",
                "name": "The Wedding Squanchers"
              },
              {
                "air_date": "April 1, 2017",
                "name": "The Rickshank Rickdemption"
              },
              {
                "air_date": "July 30, 2017",
                "name": "Rickmancing the Stone"
              },
              {
                "air_date": "August 6, 2017",
                "name": "Pickle Rick"
              },
              {
                "air_date": "August 13, 2017",
                "name": "Vindicators 3: The Return of Worldender"
              },
              {
                "air_date": "August 20, 2017",
                "name": "The Whirly Dirly Conspiracy"
              },
              {
                "air_date": "August 27, 2017",
                "name": "Rest and Ricklaxation"
              },
              {
                "air_date": "September 10, 2017",
                "name": "The Ricklantis Mixup"
              },
              {
                "air_date": "September 17, 2017",
                "name": "Morty's Mind Blowers"
              },
              {
                "air_date": "September 24, 2017",
                "name": "The ABC's of Beth"
              },
              {
                "air_date": "October 1, 2017",
                "name": "The Rickchurian Mortydate"
              },
              {
                "air_date": "November 10, 2019",
                "name": "Edge of Tomorty: Rick, Die, Rickpeat"
              },
              {
                "air_date": "November 17, 2019",
                "name": "The Old Man and the Seat"
              },
              {
                "air_date": "November 24, 2019",
                "name": "One Crew Over the Crewcoo's Morty"
              },
              {
                "air_date": "December 8, 2019",
                "name": "Claw and Hoarder: Special Ricktim's Morty"
              },
              {
                "air_date": "December 15, 2019",
                "name": "Rattlestar Ricklactica"
              },
              {
                "air_date": "May 3, 2020",
                "name": "Never Ricking Morty"
              },
              {
                "air_date": "May 10, 2020",
                "name": "Promortyus"
              },
              {
                "air_date": "May 17, 2020",
                "name": "The Vat of Acid Episode"
              },
              {
                "air_date": "May 24, 2020",
                "name": "Childrick of Mort"
              },
              {
                "air_date": "May 31, 2020",
                "name": "Star Mort: Rickturn of the Jerri"
              }
            ]
          }
        },
      },
    };
    const component = renderer.create(
      <MockedProvider mocks={[mockedData]} addTypename={false} >
        <CharacterDetailsScreen />
      </MockedProvider>,
    );

    await wait();

    console.log(component.toJSON());
    expect(() => {
      component.root.findByProps({
        children: 'Male',
      });
    });
  
  });


});