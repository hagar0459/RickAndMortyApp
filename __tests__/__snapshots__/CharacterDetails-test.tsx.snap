// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CharacterDetailsScreen Snapshot Test 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View
    style={
      Object {
        "borderRadius": 0,
        "marginBottom": 20,
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "uri": undefined,
        }
      }
      style={
        Object {
          "height": 150,
        }
      }
    />
    <View
      style={
        Object {
          "padding": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 22,
            "fontWeight": "bold",
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 16,
            "fontWeight": "600",
            "marginBottom": 8,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#787878",
            "fontSize": 16,
            "fontWeight": "400",
          }
        }
      />
    </View>
  </View>
  <Text
    style={
      Object {
        "fontSize": 16,
        "fontWeight": "600",
        "marginHorizontal": 20,
      }
    }
  >
    Episods:
  </Text>
  <RCTScrollView
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</RCTSafeAreaView>
`;
